### PR TITLE
atc(feat): expose RESOURCE_NAME to steps get and put.

### LINKED ISSUE #5089: Expose RESOURCE_NAME to `put` step
## What challenge are you facing?

Concourse exposes a set of build metadata, `BUILD_ID`, `BUILD_NAME`, etc. to `get` and `put`, but without current resource name.

When a resource is `putting`, it doesn't know who itself is, which leads to a weird situation, for example:

```yaml
resources:
- name: mr
   type: gitlab-merge-request
   ...

jobs:
- name: job1
  plan:
  - get: mr # get the merge-request metadata, and stores in files.

  - put: mr
     params:
        dir: mr
        status: running
```

The `put` on `mr` doesn't know where `mr` is, so user has to define a param `dir: mr` to tell the `put` where the resource is, which sounds and looks really weird.

## What would make this better?

If current resource name is exposed to `put`, then `put` can by default look for `/tmp/put/<current resource name>` for itself, so the above weird param `dir: mr` can be avoided.

## Are you interested in implementing this yourself?

Yes

### PR DESCRIPTION
**Pull request by evanchaoli**
Fixes #5089 .

# Changes proposed in this pull request

Just expose environment variable `RESOURCE_NAME` to `get` and `put` containers.

# Contributor Checklist
> Are the following items included as part of this PR? Please delete checkbox items that don't apply.
- [x] Unit tests
- [x] Integration tests (if applicable)
- [ ] Updated documentation (located at https://github.com/concourse/docs)
- [ ] Updated release notes (located at https://github.com/concourse/concourse/tree/master/release-notes)

# Reviewer Checklist
> This section is intended for the core maintainers only, to track review progress. Please do not
> fill out this section.
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the [BOSH](https://github.com/concourse/concourse-bosh-release) 
      and [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for the [integration tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env) (for example, if they are Garden configs that are not displayed in the `--help` text).

### CODE REVIEW

**Comment by evanchaoli** [comment]
> Do you mind moving this logic to where the step gets created in the builder instead of in the step itself: https://github.com/concourse/concourse/blob/master/atc/engine/builder/builder.go#L312

@pivotal-jwinters Not sure what you mean. Do you want to put `resource_name` in `stepMetadata`?

**Comment by vito** [comment]
Sorry but I'm pretty opposed to this change. 🙁And sorry for not replying earlier on the original issue proposing this - I've been heads-down trying to get v6.0 ready.

I don't like the idea of resources changing their behavior based on what they're named because it violates the principle that resources are equivalent as long as they're configured the same way (i.e. `type`, `source`, `params`).

Conceptually, this conflates two boundaries. In the Concourse mental model, pipelines are a resource namespace, and resources themselves are just values. To use a programming language metaphor, this is the difference between variable names and variable values. If a language's values changed when you assigned them to a different name, that would feel really strange. 😛

This principle is also important for Concourse's ability to make fairly significant optimizations.

For example, imagine a resource type author implemented `get` to do something different based on the name of the resource, resulting in different bits. Now we would have to take the resource name into account for caching, which would ultimately lead to a lot of wasted disk space. Even if we decide to ignore the name for caching, cache hits would still result in the `get` step being skipped, which would lead to kind of weird semantics where it only runs with whatever the first resource's name was.

Imagining this change also exposed the name to `check`, resource type authors may be tempted to use the resource name to identify the resource in question so that there's less to configure. But then we wouldn't be able to efficiently consolidate version history either, which would lead to a lot of extra `check`ing load across the workers.

Exposing the resource name to `put` doesn't have such significant technical ramifications, especially once we implement #3463, but I still don't think this change is worth the saved line of YAML. I think the behavior proposed in #5089 (using the resource name as a default value for a `dir` param) is a bit too magical and could mislead new users into thinking that resources have some sort of internal state or communication mechanism. It also won't help you anyway if the `get` step is renamed:

```yaml
- get: concourse
  resource: concourse-pr
- put: concourse-pr
  params: {status: pending} # breaks as this needs dir: concourse, not concourse-pr
```

In the long run, I plan to remove all of the other environment variables that we expose to resource containers. That mechanism should be replaced with proper support for notifications, which are really the only things that should have such information.

Given that the original motivation for #5089 appears to be centred around notifications, and the small change here actually has pretty significant implications, I would rather just stay focused on the long-term solution as it'll have an even larger positive impact on the original use case (you wouldn't need the `put` step *at all*). 🙏

**Comment by evanchaoli** [comment]
> And sorry for not replying earlier on the original issue proposing this - I've been heads-down trying to get v6.0 ready.

Understood!

> I don't like the idea of resources changing their behavior based on what they're named because it violates the principle that resources are equivalent as long as they're configured the same way (i.e. `type`, `source`, `params`).

I didn't mean to expose RESOURCE_NAME to `check` at all, nor `get` in #5089. Just when I wrote the code, I felt no harm to expose to `get`, so I added the logic to `get`. I'm totally ok to not expose RESOURCE_NAME to `check` and `get`.

> Exposing the resource name to `put` doesn't have such significant technical ramifications, especially once we implement #3463, but I still don't think this change is worth the saved line of YAML. I think the behavior proposed in #5089 (using the resource name as a default value for a `dir` param) is a bit too magical and could mislead new users into thinking that resources have some sort of internal state or communication mechanism. It also won't help you anyway if the `get` step is renamed:
> 
> ```yaml
> - get: concourse
>   resource: concourse-pr
> - put: concourse-pr
>   params: {status: pending} # breaks as this needs dir: concourse, not concourse-pr
> ```

I never used `resource` to rename a resource, also seldom see a pipeline on our clusters. So this example is valid, but I think it has the least impact.

We just expose a RESOURCE_NAME to `put`, it's up to resource type authors how to use it. A resource may still have a param for `put` to specify where is itself, but exposing RESOURCE_NAME  to `put` at least provides a way for default, it could be apply to most of use cases.

> In the long run, I plan to remove all of the other environment variables that we expose to resource containers. That mechanism should be replaced with proper support for notifications, which are really the only things that should have such information.
> 
> Given that the original motivation for #5089 appears to be centred around notifications, and the small change here actually has pretty significant implications, I would rather just stay focused on the long-term solution as it'll have an even larger positive impact on the original use case (you wouldn't need the `put` step _at all_). 🙏

#5089 isn't really centered around notifications. I have seen many resource types, for example, our home-brew resource types, gitlab, build-web, and so on... By removing redundant param of specifying resource itself's name help make pipeline code neat and easier to understand.

From a user's perspective, if I do get then put:
```yaml
 - get: resource-a
 - put: resource-a
```

It's totally operating the same resource object, it's hard to understand why I still need to specify a redundant param like:
```yaml
 - get: resource-a
 - put: resource-a
    param:
       dir: resource-a
```

What's reason why I think this PR will help pipeline to easier to understand. It's actually a burden for resource type authors to document a pretty explanation for such a param.

For long term, if environment variables are no longer exposed to `get` and `put`, then a number of resource types will have to rewrite, at that time, RESOURCE_NAME should not add much more burden. And what's ETA of the "long term"? I think it's worth letting current resource types enjoy RESOURCE_NAME before the "long term" comes.

In summary, I can remove the logic from `get`, but I still want to solve #5089. @vito, Are you ok with that?

**Comment by vito** [comment]
> #5089 isn't really centered around notifications.

By notifications I mean things like GitHub status, GitLab status, PR status, Slack notifications, etc. - things that update or message some external location with the status of the build. Judging by the code sample in #5089 that seems like the use case:

```yaml
  - put: mr
    params:
      dir: mr
      status: running
```

Updating PR status is pretty much the only situation where I've ended up passing a resource to itself. The whole flow is pretty brittle. For example, in our use of the `github-pr` resource we had to be careful to give the `get` step and the `put` step different names so that the `put` step's implicit `get` step doesn't clobber our original `get` step's bits (which at one point we had given `params`):

https://github.com/concourse/ci/blob/00078c07486b18613f0c5d592e2ff782a2876ab3/pipelines/prs.yml#L105-L110

To be frank, the user experience for notifications is pretty terrible at the moment. It clutters the UI, it's verbose, and it's error-prone. If notifications are the primary motivation, I don't think this change really improves the situation enough to justify increasing the possibility of "gotchas" and misleading new users. The real problem to me is that you shouldn't have to write all those `put` steps in the first place.

With notification support, we would have a way to configure a resource with a notifier. The notifier would run for each state change to a build that uses the resource. The long-term plan is to allow Prototypes to implement a notifier interface, but if we want to take a shortcut we could just implement a few notifier types natively and mark them as 'experimental' like we did with `var_sources`. A basic experimental implementation wouldn't take long to implement and would have a much larger impact.

I'd be happy to help!

(P.S. Honestly it's kind of uncanny how your PRs keep straddling the line of things I haven't yet completed an RFC for but have put vague descriptions on the roadmap. 😅 This is the fourth one in a row - `set_pipeline` step, `var_sources`, `load_var` step, and now notifications. Sorry. You're just about driving v10 from the outside at this point. I owe you many beers.)

**Comment by evanchaoli** [comment]
> (P.S. Honestly it's kind of uncanny how your PRs keep straddling the line of things I haven't yet completed an RFC for but have put vague descriptions on the roadmap. 😅 This is the fourth one in a row - `set_pipeline` step, `var_sources`, `load_var` step, and now notifications. Sorry. You're just about driving v10 from the outside at this point. I owe you many beers.)

Sorry about that. I'm going to stop after #5034. Also hopefully #5091 be merged to 6.0 as well that is a small feature, but I believe it benefits clusters.

**Comment by vito** [comment]
> Sorry about that. I'm going to stop after #5034. Also hopefully #5091 be merged to 6.0 as well that is a small feature, but I believe it benefits clusters.

Haha, I'm the one that's sorry. I didn't mean that to discourage you. I just felt bad about always having some farther-off thing to promise instead of just merging in your contributions. I'm actually really grateful for your work and enjoy collaborating towards a common goal, I was just worried that this pattern is becoming frustrating for you. 😟

Cheers!

**Comment by evanchaoli** [comment]
@vito TBH, I don't have a clear view of "v10" that you mentioned, I'm just driven by my own requirements, where #5034 is the last big feature in short term, that's reason why I said I'm going to stop.

For this PR, I didn't associate it to `notification` at all, I thought it's just an innocent small feature, as I mentioned, we do have some resource types not related to notification at all. Letting a `put` know its own resource name sounds reasonable. I don't think this feature conflicts to `notification`. From my understanding, notification mainly solves the implied get problem. Even if `notification` is released, it requires resource types authors to support the new interface, and pipeline authors to update pipelines, which takes time and effort, so some `put` might be there forever.

I won't argue more, though I'm still not convinced.

**Comment by evanchaoli** [comment]
Closing as this PR is not going to be accepted.
